vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,920 @@
+            lun,
+            length: storvsp_protocol::SCSI_REQUEST_LEN_V2 as u16,
+            cdb_length: size_of::<scsi_defs::Cdb10>() as u8,
+            data_transfer_length: byte_len as u32,
+            ..FromZeros::new_zeroed()
+        };
+
+        scsi_req.payload[0..10].copy_from_slice(cdb.as_bytes());
+        scsi_req
+    }
+
+    // This function assumes the sector size is 512.
+    fn generate_read_packet(
+        target_id: u8,
+        path_id: u8,
+        lun: u8,
+        block: u32,
+        byte_len: usize,
+    ) -> storvsp_protocol::ScsiRequest {
+        let cdb = scsi_defs::Cdb10 {
+            operation_code: ScsiOp::READ,
+            logical_block: block.into(),
+            transfer_blocks: ((byte_len / 512) as u16).into(),
+            ..FromZeros::new_zeroed()
+        };
+
+        let mut scsi_req = storvsp_protocol::ScsiRequest {
+            target_id,
+            path_id,
+            lun,
+            length: storvsp_protocol::SCSI_REQUEST_LEN_V2 as u16,
+            cdb_length: size_of::<scsi_defs::Cdb10>() as u8,
+            data_transfer_length: byte_len as u32,
+            ..FromZeros::new_zeroed()
+        };
+
+        scsi_req.payload[0..10].copy_from_slice(cdb.as_bytes());
+        scsi_req
+    }
+
+    #[async_test]
+    async fn test_negotiation(driver: DefaultDriver) {
+        let (guest, host) = connected_async_channels(16 * 1024);
+        let host_queue = Queue::new(host).unwrap();
+        let test_guest_mem = GuestMemory::allocate(16384);
+
+        let storvsp = TestStorvspWorker::start(
+            driver.clone(),
+            test_guest_mem.clone(),
+            host_queue,
+            Vec::new(),
+        );
+        let mut storvsc = TestStorvscWorker::new();
+        storvsc.start(driver.clone(), guest);
+
+        let mut timer = PolledTimer::new(&driver);
+        timer.sleep(time::Duration::from_secs(1)).await;
+
+        storvsc.stop().await;
+        assert!(storvsc.get_mut().has_negotiated);
+
+        storvsc.teardown().await;
+        storvsp.teardown().await;
+    }
+
+    #[async_test]
+    async fn test_request_response(driver: DefaultDriver) {
+        let (guest, host) = connected_async_channels(16 * 1024);
+        let host_queue = Queue::new(host).unwrap();
+        let test_guest_mem = GuestMemory::allocate(16384);
+
+        let storvsp = TestStorvspWorker::start(
+            driver.clone(),
+            test_guest_mem.clone(),
+            host_queue,
+            Vec::new(),
+        );
+        let mut storvsc = TestStorvscWorker::new();
+        storvsc.start(driver.clone(), guest);
+
+        let mut timer = PolledTimer::new(&driver);
+        timer.sleep(time::Duration::from_secs(1)).await;
+
+        storvsc.stop().await;
+        assert!(storvsc.get_mut().has_negotiated);
+        storvsc.resume().await;
+
+        // Send SCSI write request
+        let write_buf = [7u8; 4096];
+        test_guest_mem.write_at(4096, &write_buf).unwrap();
```

Comment:
This looks like the same code that is in `storage_tests/tests/storvsc.rs`. What's the difference between them?

Reply:
I see. These tests don't have a "host" side controller or disk, whereas the ones in `storage_tests` do. I'm not sure what we get by having the tests with a backend disk. Presumably it sets us up for some more interesting tests in the future? Was that your intention?